File: C++/51. N-Queens.cpp
/*
    Problem Statement:
    Given an integer n, return all distinct solutions to the n-queens puzzle.

    Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a queen and an empty space respectively.
    
    Problem Link:
    https://leetcode.com/problems/n-queens/

    Approach:
    Basic Recursive solution.
*/

class Solution {
public:
    vector<string> t;
    vector<vector<int>> correct;
    int N;
    void doIt(vector<int>&x,int col){
        if(col==N){
            correct.push_back(x);
            return;
        }
        set<int> cnt,cnt1,cnt2;
        for(int i = 0;i<col;i++) cnt.insert(x[i]);
        for(int i = 0;i<col;i++) cnt1.insert(i+x[i]);
        for(int i = 0;i<col;i++) cnt2.insert(i-x[i]);
        for(int i = 0;i<N;i++){
            if(cnt.find(i)==cnt.end() && cnt1.find(i+col)==cnt1.end() && cnt2.find(col-i)==cnt2.end()){
                x[col] = i;
                doIt(x,col+1);
            }
        }
        return;
    }
    
    vector<vector<string>> solveNQueens(int n) {
        string temp = "";
        N = n;
        for(int i = 0;i<n;i++) temp+=".";
        for(int i = 0;i<n;i++) t.push_back(temp);
        vector<int> x(n,-1);
        doIt(x,0);    
        vector<vector<string>> ans;
        for(auto i: correct){
            for(auto j: i) cout<<j<<" ";
            cout<<endl;
        }
        for(auto i: correct){
            vector<string> cp = t;
            for(int j=0;j<n;j++){
                cp[j][i[j]]='Q';
            }
            ans.push_back(cp);
        }
        return ans;
    }
};

File: Python/51. N-Queens.py
""" Problem - 51. N-Queens

Problem statement -
The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.

Given an integer n, return all distinct solutions to the n-queens puzzle.

Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a
queen and an empty space respectively.


Example :

Input: 4
Output: [
 [".Q..",  // Solution 1
  "...Q",
  "Q...",
  "..Q."],

 ["..Q.",  // Solution 2
  "Q...",
  "...Q",
  ".Q.."]
]
Explanation: There exist two distinct solutions to the 4-queens puzzle as shown above.



//Solution
In this problem, we can go row by row, and in each position, we need to check if the column, the 45° diagonal and the 135° diagonal had a queen before.

"""


class Solution(object):
    def solveNQueens(self, n):
        def dfs(r):
            if r == n:
                res.append([''.join(row) for row in b])
                return
            for c in xrange(n):
                if isValid(r, c):
                    b[r][c] = 'Q'
                    dfs(r + 1)  # fill row by row
                    b[r][c] = '.'

        def isValid(r, c):
            for i in xrange(r):
                for j in xrange(n):
                    if b[i][j] == 'Q' and (c == j or  # same column
                                           i + j == r + c or  # 45 degree line
                                           i - j == r - c):  # 135 degree line
                        return False
            return True

        b = [['.'] * n for _ in xrange(n)]
        res = []
        dfs(0)  # start from row 0
        return res
